### PR TITLE
Allow to toggle sidebar and always show the toggle

### PR DESCRIPTION
**Pull request by ma12-co**
- [x] Move the `AppNavigationToggle` in the `AppNavigation`
- [x] Move from #app-navigation IDs to have independent styling from server
- [x] Move `AppNavigation` open state IN the `AppNavigation`
- [x] Create a global `toggle-navigation` to emit with th event bus for any app to open the navigation (was originally the start of this pair programming, we ended up fixing far too much ^^')
- [x] Allow to toggle the navigation even when in desktop mode
- [x] Changed the animation/css from translate and shifting the `AppContent` to the left to using proper flex layout and not weird absolute positioning anymore :tada:
- [x] In mobile, opening the navigation does NOT shift the content, it just hovers it now. It avoid changing the whole layout and improve the rendering performances
- [x] Document the new global event (cc @ChristophWurst)
- [x] Emit event at first load

### QA
- [x] Swiping
- [x] Toggling navigation
- [x] Scrolling navigation
- [x] Not being able to scroll left and see the hidden navigation
- [x] Settings

fix https://github.com/nextcloud/server/issues/15831
___
@ma12-co, @skjnldsv & @ChristophWurst on a nice beer programming session on a sweet Friday afternoon :beers:

### CODE REVIEW

**Comment by skjnldsv** [comment]
Capture d’écran_2020-04-03_20-00-17

**Comment by skjnldsv** [comment]
image

**Comment by korelstar** [comment]
Another thing: if I want the app-navigation to be closed by default (even in non-mobile mode), then I would have to emit an event using the event-bus on app initialization? Wouldn't it be easier to provide a prop for the default state?

**Comment by skjnldsv** [comment]
> Another thing: if I want the app-navigation to be closed by default (even in non-mobile mode), then I would have to emit an event using the event-bus on app initialization? Wouldn't it be easier to provide a prop for the default state?

I would be up for a default state, but does that make sense in any app? We still want to show the Navigation by default in almost all of them, no?

**Comment by korelstar** [comment]
Yes, there is a need. Please see https://github.com/nextcloud/nextcloud-vue/issues/791 (and the linked issues) for a discussion on this.

**Comment by skjnldsv** [comment]
> Yes, there is a need. Please see #791 (and the linked issues) for a discussion on this.

Commenting on #791 as there seems to be a mix up with various issues since it was last discussed :p

**Comment by jancborchardt** [comment]
Very nice!


>* `optional` remember state ?

What is meant by this? If we have remembering, it should be per-app, and it should _always_ work, not giving app developers a way to disable it.

**Comment by skjnldsv** [comment]
> What is meant by this? If we have remembering, it should be per-app, and it should _always_ work, not giving app developers a way to disable it.

It means it can be in a separate pr to keep it clean.

**Comment by georgehrke** [comment]
@ma12-co Thanks! This also fixes an issue we had in the calendar, where the toggle was shown, even though there is no app-navigation. ❤️

**Comment by ma12-co** [comment]
you should thank @skjnldsv more than me :)

**Comment by skjnldsv** [comment]
> We should make sure that the spec for "New message" and "New contact" buttons as discussed in [nextcloud/server#16934 (comment)](https://github.com/nextcloud/server/issues/16934#issuecomment-542201187) is implemented before merging this. Otherwise the issue of **"Mobile menu toggle for navigation overlaps with content (especially in Mail & Contacts)"** will not only exist on mobile but on desktop too.

I don't know actually.
I would rather keep it simple here, otherwise it's harder to review.
Also because technically they're not the same and require us to call and decide of the proper implementation.
This doesn't create regressions technically. Let's fix the remaining issues in OP and merge it :)

File: src/components/AppContent/AppContent.vue [hunk 1]
```diff
@@ -30,13 +30,40 @@
...
 /**
  * App content container to be used for the main content of your app
  *
  */
 export default {
 	name: 'AppContent',
+
+	mounted() {
+		this.mc = new Hammer(this.$el, { cssProps: { userSelect: 'text' } })
+		this.mc.on('swipeleft swiperight', e => {
```
Comment (juliushaertl) [inline src/components/AppContent/AppContent.vue:45]: Can we add a toggle for this, as for deck where we have horizontal scrolling we had quite some issues in the past when reordering cards with drag-and-drop caused the sidebar drag to trigger.

File: src/components/AppContent/AppContent.vue [hunk 2]
```diff
@@ -21,78 +23,59 @@
...
-	data() {
-		return {
-			// closed by default on mobile mode
-			opened: false,
-		}
-	},
+
 	mounted() {
 		this.mc = new Hammer(this.$el, { cssProps: { userSelect: 'text' } })
 		this.mc.on('swipeleft swiperight', e => {
```
Comment (ChristophWurst) [inline src/components/AppContent/AppContent.vue:45]:
```suggestion
		this.mc.on('swipeleft swiperight', this.handleSwipe)
```

and `this.mc.off(...) in unmounted

File: src/components/AppContent/AppContent.vue (hunk 1 above)
Comment (ma12-co) [reply src/components/AppContent/AppContent.vue]: Done!

Comment (juliushaertl) [reply src/components/AppContent/AppContent.vue]: Awesome :rocket:

File: src/components/AppContent/AppContent.vue (hunk 2 above)
Comment (ma12-co) [reply src/components/AppContent/AppContent.vue]: Done!

**Comment by ma12-co** [comment]
@skjnldsv I guess those conflicts are because of #1002, what should I do here?

**Comment by skjnldsv** [comment]
> @skjnldsv I guess those conflicts are because of #1002, what should I do here?

Git rebase with strategy `ours` to keep the current master files and manually `npm i @nextcloud/event-bus` again

**Comment by skjnldsv** [comment]
Docs fails https://app.netlify.com/sites/nextcloud-vue-components/deploys/5e8e1c5c0203164eaddaf323

**Comment by skjnldsv** [comment]
WORKED

**Comment by ma12-co** [comment]
> WORKED
Oh nice!

**Comment by jancborchardt** [comment]
> > We should make sure that the spec for "New message" and "New contact" buttons as discussed in [nextcloud/server#16934 (comment)](https://github.com/nextcloud/server/issues/16934#issuecomment-542201187) is implemented before merging this. Otherwise the issue of **"Mobile menu toggle for navigation overlaps with content (especially in Mail & Contacts)"** will not only exist on mobile but on desktop too.
> 
> I don't know actually.
> I would rather keep it simple here, otherwise it's harder to review.
> Also because technically they're not the same and require us to call and decide of the proper implementation.
> This doesn't create regressions technically. Let's fix the remaining issues in OP and merge it :)

Fine with me as well, as then it will make it obvious that it will have to be fixed. :D

**Pull request merged**